File: template/.github/scripts/validate-selected-contract-scalar-placeholders.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative "lib/profile_contracts"

SKILL_PATH = "SKILL.md"
ROUTING_PATH = "INTERFACES.md"
RUNTIME_PATH = "RUNTIME.md"
WEB_PATH = "WEB_INTERFACE.md"
RUNTIME_REQUIRED_PROFILES = %w[packaged-cli mcp-enabled browser-interface headless-service].freeze

begin
  selection = ProfileContracts::ProfileSelection.load(SKILL_PATH)
rescue ProfileContracts::ParseError => error
  warn error.message
  exit 1
end

if selection.template_scaffold?
  puts "Selected-contract scalar placeholder validation is not activated for the template scaffold."
  exit 0
end

policy = ProfileContracts::ValuePolicy
errors = []

scan_scalar_values = lambda do |path, document, context = nil|
  document.each_scalar do |entry|
    next unless policy.unresolved_scalar?(entry.value)

    location = context ? "#{path} #{context}" : "#{path}:#{entry.line_number}"
    value = policy.strip_backticks(entry.value)

    case entry.kind
    when :field
      errors << "#{location} '#{entry.label}:' must not use unresolved scalar placeholder #{value.inspect}."
    when :table
      errors << "#{location} table value must not use unresolved scalar placeholder #{value.inspect}."
    else
      errors << "#{location} must not contain standalone unresolved scalar placeholder #{value.inspect}."
    end
  end
end

skill_document = ProfileContracts::MarkdownDocument.read(SKILL_PATH)
selected_profiles = selection.profiles

# Every concrete skill must be operational regardless of whether it retains an
# implementation runtime. Scan the entire SKILL document so instruction-only,
# knowledge, asset, and script profiles cannot leave TBD-style placeholders in
# their purpose, trigger, workflow, output, validation, or safety sections.
scan_scalar_values.call(SKILL_PATH, skill_document)

routing_selected = (selected_profiles & %w[packaged-cli mcp-enabled]).any?
runtime_required = (selected_profiles & RUNTIME_REQUIRED_PROFILES).any?
runtime_retained = runtime_required || (selection.selected?("script-assisted") && File.file?(RUNTIME_PATH))

if runtime_retained
  ["Canonical command", "Working directory"].each do |label|
    skill_document.summary_values(label).each do |value|
      if policy.unresolved_scalar?(value)
        errors << "#{SKILL_PATH} '#{label}:' must not use unresolved scalar placeholder #{value.inspect}."
      end
    end
  end
end

if routing_selected
  ["Preferred agent route", "Detailed interface contract"].each do |label|
    skill_document.summary_values(label).each do |value|
      if policy.unresolved_scalar?(value)
        errors << "#{SKILL_PATH} '#{label}:' must not use unresolved scalar placeholder #{value.inspect}."
      end
    end
  end
end

selected_contracts = []
selected_contracts << ROUTING_PATH if routing_selected
selected_contracts << "CLI_INTERFACE.md" if selection.selected?("packaged-cli")
selected_contracts << "MCP_INTERFACE.md" if selection.selected?("mcp-enabled")
selected_contracts << WEB_PATH if selection.selected?("browser-interface")

selected_contracts.each do |path|
  unless File.file?(path)
    errors << "Selected profile requires contract file: #{path}"
    next
  end

  scan_scalar_values.call(path, ProfileContracts::MarkdownDocument.read(path))
end

if runtime_required && !File.file?(RUNTIME_PATH)
  errors << "Selected runtime-backed profile requires contract file: #{RUNTIME_PATH}"
elsif runtime_retained
  runtime = ProfileContracts::MarkdownDocument.read(RUNTIME_PATH)
  runtime_headings = [
    "## Status",
    "## Primary implementation",
    "### Shared development commands",
    "## Distribution",
    "## Environment and configuration",
    "## Decision rationale"
  ]
  runtime_headings << "### Packaged CLI commands" if selection.selected?("packaged-cli")
  if selection.selected?("mcp-enabled")
    runtime_headings.concat([
      "### MCP commands",
      "## MCP protocol support",
      "### stdio variant",
      "### Streamable HTTP variant",
      "### Bundled ad hoc MCP tool client"
    ])
  end
  if selection.selected?("browser-interface")
    runtime_headings.concat([
      "### Browser-interface commands",
      "## Optional human verification Web interface deployment"
    ])
  end
  if selection.selected?("headless-service")
    runtime_headings.concat([
      "### Headless-service commands",
      "## Headless service deployment"
    ])
  end

  runtime_headings.uniq.each do |heading|
    section = runtime.section(heading)
    next unless section

    scan_scalar_values.call(
      RUNTIME_PATH,
      ProfileContracts::MarkdownDocument.new(section, path: RUNTIME_PATH),
      heading
    )
  end
end

unless errors.empty?
  errors.uniq.each { |error| warn error }
  exit 1
end

puts "Concrete SKILL, selected routing/interface, and runtime scalar values contain no unresolved placeholders."

File: template/.github/scripts/validate_selected_contract_scalar_placeholders.py
#!/usr/bin/env python3
"""Reject unresolved scalar placeholders in selected concrete contracts."""

from __future__ import annotations

import sys
from pathlib import Path

from lib.profile_contracts import MarkdownDocument, ParseError, ProfileSelection, ValuePolicy


SKILL_PATH = Path("SKILL.md")
ROUTING_PATH = Path("INTERFACES.md")
RUNTIME_PATH = Path("RUNTIME.md")
WEB_PATH = Path("WEB_INTERFACE.md")
RUNTIME_REQUIRED_PROFILES = {
    "packaged-cli",
    "mcp-enabled",
    "browser-interface",
    "headless-service",
}


def run() -> int:
    try:
        selection = ProfileSelection.load(SKILL_PATH)
    except (ParseError, OSError) as exc:
        print(exc, file=sys.stderr)
        return 1

    if selection.template_scaffold():
        print(
            "Selected-contract scalar placeholder validation is not activated "
            "for the template scaffold."
        )
        return 0

    errors: list[str] = []

    def scan_scalar_values(
        path: Path,
        document: MarkdownDocument,
        context: str | None = None,
        line_offset: int = 0,
    ) -> None:
        for entry in document.each_scalar():
            if not ValuePolicy.unresolved_scalar(entry.value):
                continue

            location = f"{path}:{line_offset + entry.line_number}"
            if context:
                location += f" ({context})"
            value = ValuePolicy.strip_backticks(entry.value)
            rendered = repr(value)

            if entry.kind == "field":
                errors.append(
                    f"{location} '{entry.label}:' must not use unresolved "
                    f"scalar placeholder {rendered}."
                )
            elif entry.kind == "table":
                errors.append(
                    f"{location} table value must not use unresolved "
                    f"scalar placeholder {rendered}."
                )
            else:
                errors.append(
                    f"{location} must not contain standalone unresolved "
                    f"scalar placeholder {rendered}."
                )

    skill_document = MarkdownDocument.read(SKILL_PATH)
    selected_profiles = set(selection.profiles)

    scan_scalar_values(SKILL_PATH, skill_document)

    routing_selected = bool(
        selected_profiles.intersection({"packaged-cli", "mcp-enabled"})
    )
    runtime_required = bool(
        selected_profiles.intersection(RUNTIME_REQUIRED_PROFILES)
    )
    runtime_retained = runtime_required or (
        selection.selected("script-assisted") and RUNTIME_PATH.is_file()
    )

    if runtime_retained:
        for label in ("Canonical command", "Working directory"):
            for value in skill_document.summary_values(label):
                if ValuePolicy.unresolved_scalar(value):
                    errors.append(
                        f"{SKILL_PATH} '{label}:' must not use unresolved "
                        f"scalar placeholder {value!r}."
                    )

    if routing_selected:
        for label in ("Preferred agent route", "Detailed interface contract"):
            for value in skill_document.summary_values(label):
                if ValuePolicy.unresolved_scalar(value):
                    errors.append(
                        f"{SKILL_PATH} '{label}:' must not use unresolved "
                        f"scalar placeholder {value!r}."
                    )

    selected_contracts: list[Path] = []
    if routing_selected:
        selected_contracts.append(ROUTING_PATH)
    if selection.selected("packaged-cli"):
        selected_contracts.append(Path("CLI_INTERFACE.md"))
    if selection.selected("mcp-enabled"):
        selected_contracts.append(Path("MCP_INTERFACE.md"))
    if selection.selected("browser-interface"):
        selected_contracts.append(WEB_PATH)

    for path in selected_contracts:
        if not path.is_file():
            errors.append(f"Selected profile requires contract file: {path}")
            continue
        scan_scalar_values(path, MarkdownDocument.read(path))

    if runtime_required and not RUNTIME_PATH.is_file():
        errors.append(
            f"Selected runtime-backed profile requires contract file: {RUNTIME_PATH}"
        )
    elif runtime_retained:
        runtime = MarkdownDocument.read(RUNTIME_PATH)
        runtime_headings = [
            "## Status",
            "## Primary implementation",
            "### Shared development commands",
            "## Distribution",
            "## Environment and configuration",
            "## Decision rationale",
        ]
        if selection.selected("packaged-cli"):
            runtime_headings.append("### Packaged CLI commands")
        if selection.selected("mcp-enabled"):
            runtime_headings.extend(
                [
                    "### MCP commands",
                    "## MCP protocol support",
                    "### stdio variant",
                    "### Streamable HTTP variant",
                    "### Bundled ad hoc MCP tool client",
                ]
            )
        if selection.selected("browser-interface"):
            runtime_headings.extend(
                [
                    "### Browser-interface commands",
                    "## Optional human verification Web interface deployment",
                ]
            )
        if selection.selected("headless-service"):
            runtime_headings.extend(
                [
                    "### Headless-service commands",
                    "## Headless service deployment",
                ]
            )

        for heading in dict.fromkeys(runtime_headings):
            section = runtime.section(heading)
            if section is None:
                continue
            heading_line_number = next(
                (
                    index
                    for index, line in enumerate(runtime.lines, start=1)
                    if line.rstrip() == heading
                ),
                0,
            )
            scan_scalar_values(
                RUNTIME_PATH,
                MarkdownDocument(section, path=RUNTIME_PATH),
                heading,
                line_offset=heading_line_number,
            )

    if errors:
        for error in dict.fromkeys(errors):
            print(error, file=sys.stderr)
        return 1

    print(
        "Concrete SKILL, selected routing/interface, and runtime scalar values "
        "contain no unresolved placeholders."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(run())
